cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/replication/ReplicationThreadpoolTest.java

```diff
@@ -0,0 +1,258 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ReplicationThreadpoolTest {
+
+    /**
+     * This test validates that the type of ThreadPoolExecutor we use for GetRevisionTaskThreaded
+     * behaves as we expect in terms of generation and expiry of threads.
+     * @throws Exception
+     */
+    @Test
+    public void threadpoolTest() throws Exception {
+
+        final CountDownLatch latch = new CountDownLatch(1);
+        int threads = 4;
+        ThreadPoolExecutor tpe = new ThreadPoolExecutor(threads, threads, 1,
+                TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>());
+        // Allowing core threads to timeout means we don't keep threads in memory except when
+        // tasks are running or for the configured amount of time afterwards.
+        tpe.allowCoreThreadTimeOut(true);
+
+        List<Future<?>> futures = new ArrayList<Future<?>>();
+
+        for (int i = 1; i < threads; i++) {
+            futures.add(tpe.submit(new LatchedRunnable(latch)));
+            Assert.assertEquals("The threadpool count should increase with tasks up to threads",
+                    i, tpe.getPoolSize());
+        }
+        futures.add(tpe.submit(new LatchedRunnable(latch)));
+        Assert.assertEquals("The threadpool count should be capped", threads, tpe.getPoolSize());
+
+        // Now countdown the latch to release the threads
+        latch.countDown();
+        // Ensure they are all completed
+        for (Future<?> f : futures) {
+            f.get();
+        }
+
+        // Sleep to allow the threads to expire (1 second expiry, but allow 2)
+        TimeUnit.SECONDS.sleep(2);
+        // Validate no threads remain
+        Assert.assertEquals("The threadpool should be empty", 0, tpe.getPoolSize());
+    }
+
+    /*
+     * Simple test to execute a method which adds numbers over a large number of threads
+     */
+    @Test
+    public void addServiceTest() {
+        Queue<AddRequest> requests = new LinkedBlockingQueue<AddRequest>();
+        int nRequests = 10000;
+        for (int i=0; i < nRequests; i++) {
+            AddRequest request = new AddRequest();
+            request.x = i;
+            request.y = i*2;
+            requests.add(request);
+        }
+        int threads = 50;
+        final AtomicInteger nResponses = new AtomicInteger();
+        QueuingExecutorCompletionService<AddRequest, AddResponse> service =
+                new QueuingExecutorCompletionService<AddRequest, AddResponse>(Executors.newFixedThreadPool(threads), requests, threads) {
```

Comment:
This statement requires you to scroll when reviewing.  It is already spread across two lines, if possible can you spread it across three for easier review in the future?